dmoj/judge.py

```diff
@@ -83,202 +88,162 @@ def update_problems(self):
-                        )
-                        if not is_sc
-                        else ''
-                    )
-                    case_padding = '  ' * in_batch
-                    report(
-                        ansi_style('%sTest case %2d %-3s %s' % (case_padding, case_number, colored_codes[0], case_info))
-                    )
-
-                    self.packet_manager.test_case_status_packet(case_number, result)
-
-                    case_number += 1
-        except TerminateGrading:
-            self.packet_manager.submission_terminated_packet()
-            report(ansi_style('#ansi[Forcefully terminating grading. Temporary files may not be deleted.](red|bold)'))
-            pass
-        else:
-            self.packet_manager.grading_end_packet()
-
     def begin_grading(
         self,
-        id,
+        submission_id,
         problem_id,
         language,
         source,
         time_limit,
         memory_limit,
         short_circuit,
         meta,
-        report=print,
+        report=logger.info,
         blocking=False,
     ):
-        self.current_submission_id = id
-
-        def grading_cleanup_wrapper():
+        self.current_submission_id = submission_id
+        ipc_ready_signal = threading.Event()
+        sub_judge_thread = threading.Thread(
+            target=self._run_grading_thread,
+            args=(
+                ipc_ready_signal,
+                submission_id,
+                problem_id,
+                language,
+                source,
+                time_limit,
+                memory_limit,
+                short_circuit,
+                meta,
+                report,
+            ),
+            daemon=True,
+        )
+        sub_judge_thread.start()
+
+        ipc_ready_signal.wait()
+        if blocking:
+            sub_judge_thread.join()
+
+    def _run_grading_thread(
+        self,
+        ipc_ready_signal,
+        submission_id,
+        problem_id,
+        language,
+        source,
+        time_limit,
+        memory_limit,
+        short_circuit,
+        meta,
+        report,
+    ):
+        try:
             report(
                 ansi_style(
-                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...' % (problem_id, id, language)
+                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...'
+                    % (problem_id, submission_id, language)
                 )
             )
 
-            try:
-                problem = Problem(problem_id, time_limit, memory_limit, meta)
-                self._block_and_grade(problem, language, source, short_circuit, report=report)
-            except Exception:
-                self.log_internal_error()
+            self.current_judge_process = SubmissionJudgeProcess(
+                submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta
+            )
+
+            ipc_ready_signal.set()
```

Comment:
Does this help? The actual process may not be created here, so still a race.

Reply:
I think by this point in time we *have* at least forked, so the pipe will block rather than fail, which is fine for us.

Reply:
Wrote a fancier startup where judge waits for subjudge to say "HELLO" via IPC before setting the signal.